clarity/src/vm/functions/principals.rs

```diff
@@ -38,13 +69,259 @@ pub fn special_is_standard(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    check_argument_count(1, args)?;
+    runtime_cost(ClarityCostFunction::Unimplemented, env, 0)?;
+
+    let principal = eval(&args[0], env, context)?;
+
+    let (version_byte, hash_bytes, name_opt) = match principal {
+        Value::Principal(PrincipalData::Standard(StandardPrincipalData(version, bytes))) => {
+            (version, bytes, None)
+        }
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier { issuer, name })) => {
+            (issuer.0, issuer.1, Some(name))
+        }
+        _ => {
+            return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, principal).into())
+        }
+    };
+
+    // `version_byte_is_valid` determines whether the returned `Response` is through the success
+    // channel or the error channel.
+    let version_byte_is_valid = version_matches_current_network(version_byte, env.global_context);
+
+    let tuple = create_principal_parse_tuple(version_byte, &hash_bytes, name_opt);
+    Ok(Value::Response(ResponseData {
+        committed: version_byte_is_valid,
+        data: Box::new(tuple),
+    }))
+}
+
+pub fn special_principal_construct(
+    args: &[SymbolicExpression],
+    env: &mut Environment,
+    context: &LocalContext,
+) -> Result<Value> {
+    check_arguments_at_least(2, args)?;
+    check_arguments_at_most(3, args)?;
+    runtime_cost(ClarityCostFunction::Unimplemented, env, 0)?;
+
+    let version = eval(&args[0], env, context)?;
+    let hash_bytes = eval(&args[1], env, context)?;
+    let name_opt = if args.len() > 2 {
+        Some(eval(&args[2], env, context)?)
+    } else {
+        None
+    };
+
+    // Check the version byte.
+    let verified_version = match version {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => data,
+        _ => {
+            return {
+                // This is an aborting error because this should have been caught in analysis pass.
+                Err(CheckErrors::TypeValueError(BUFF_1.clone(), version).into())
+            };
+        }
+    };
+
+    // This is an aborting error because this should have been caught in analysis pass.
+    if verified_version.len() > 1 {
+        return Err(CheckErrors::TypeValueError(BUFF_1.clone(), version).into());
+    }
+
+    // If the version byte buffer has 0 bytes, this is a recoverable error, because it wasn't the
+    // job of the type system.
+    if verified_version.len() < 1 {
+        // do some kind of error
+        return Ok(create_principal_true_error_response(1));
+    }
+
+    // Assume: verified_version.len() == 1
+    let version_byte = (*verified_version)[0];
+
+    // If the version byte is >= 32, this is a runtime error, because it wasn't the job of the
+    // type system.  This is a requirement for c32check encoding.
+    if version_byte >= 32 {
+        return Ok(create_principal_true_error_response(1));
```

Comment:
The error codes should have consts (this method uses a couple different error codes)

Reply:
Fixed